tools/cli/table.go

```diff
@@ -0,0 +1,207 @@
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package cli
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"reflect"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/olekukonko/tablewriter"
+
+	"github.com/uber/cadence/common/types"
+)
+
+// TableOptions allows passing optional flags for altering rendered table
+type TableOptions struct {
+	// OptionalColumns may contain column header names which can be hidden
+	OptionalColumns map[string]bool
+
+	// Border specified whether to render table border
+	Border bool
+
+	// NoColor will not use color characters while printing table
+	NoColor bool
+
+	// PrintRawTime will print time as int64 unix nanos
+	PrintRawTime bool
+	// PrintDateTime will print both date & time
+	PrintDateTime bool
+}
+
+// RenderTable is generic function for rendering a slice of structs as a table
+func RenderTable(w io.Writer, slice interface{}, opts TableOptions) {
+	sliceValue := reflect.ValueOf(slice)
+	if sliceValue.Kind() != reflect.Slice {
+		panic("table must be a slice")
+	}
+
+	// No elements - nothing to render
+	if sliceValue.Len() == 0 {
+		return
+	}
+
+	firstElem := sliceValue.Index(0)
+	if firstElem.Kind() != reflect.Struct {
+		panic("slice element must be a struct")
+	}
+
+	table := tablewriter.NewWriter(w)
+	table.SetBorder(opts.Border)
+	table.SetColumnSeparator("|")
+	table.SetHeaderLine(false)
+
+	for r := 0; r < sliceValue.Len(); r++ {
+		var row []string
+		var headers []string
+		var colors []tablewriter.Colors
+
+		elem := sliceValue.Index(r)
+		for f := 0; f < elem.NumField(); f++ {
+			tag := elem.Type().Field(f).Tag
+
+			header := columnHeader(tag, opts)
+			if header == "" {
+				continue
+			}
+			if r == 0 {
+				headers = append(headers, header)
+				colors = append(colors, tableHeaderBlue)
+			}
+
+			row = append(row, formatValue(elem.Field(f).Interface(), opts, tag))
+		}
+		if r == 0 {
+			table.SetHeader(headers)
+			if !opts.NoColor {
```

Comment:
NIT:
for better readability, I would do the opposite.
If `NoColor` is set, then remove colors.

Reply:
Agree, updated.